test/common/router/router_test.cc

```diff
@@ -951,4 +951,70 @@ TEST_F(RouterTest, CanaryStatusFalse) {
   EXPECT_EQ(0U, cm_.cluster_.info_->stats_store_.counter("canary.upstream_rq_200").value());
 }
 
+TEST_F(RouterTest, AutoHostRewriteEnabled) {
+  NiceMock<Http::MockStreamEncoder> encoder;
+  std::string dns_host{"scooby.doo"};
+  std::string req_host{"foo.bar.com"};
+
+  Http::TestHeaderMapImpl incoming_headers;
+  HttpTestUtility::addDefaultHeaders(incoming_headers);
+  incoming_headers.Host()->value(req_host);
+
+  Http::TestHeaderMapImpl outgoing_headers;
+  HttpTestUtility::addDefaultHeaders(outgoing_headers);
+  outgoing_headers.Host()->value(dns_host);
+
+  EXPECT_CALL(callbacks_.route_->route_entry_, timeout())
+      .WillOnce(Return(std::chrono::milliseconds(0)));
+
+  EXPECT_CALL(cm_.conn_pool_, newStream(_, _))
+      .WillOnce(Invoke([&](Http::StreamDecoder&, Http::ConnectionPool::Callbacks& callbacks)
+                           -> Http::ConnectionPool::Cancellable* {
+                             callbacks.onPoolReady(encoder, cm_.conn_pool_.host_);
+                             return nullptr;
+                           }));
+
+  EXPECT_CALL(encoder, encodeHeaders(HeaderMapEqualRef(&outgoing_headers), true))
+      .WillOnce(Invoke([&](const Http::HeaderMap&, bool) -> void {
+        encoder.stream_.resetStream(Http::StreamResetReason::RemoteReset);
+      }));
+
+  EXPECT_CALL(callbacks_.request_info_, onUpstreamHostSelected(_))
+      .WillOnce(Invoke([&](const Upstream::HostDescriptionPtr host)
+                           -> void { EXPECT_EQ(host_address_, host->address()); }));
+  EXPECT_CALL(callbacks_.route_->route_entry_, autoHostRewrite()).WillOnce(Return(true));
+  EXPECT_CALL(*cm_.conn_pool_.host_, hostname()).WillRepeatedly(ReturnRef(dns_host));
```

Comment:
It would be better if you made hostname_ a variable in the host mock which defaults to "" and then set up an ON_CALL for it. Then your tests just need to change the value of hostname_ inside the mock both here and below. It's a little less error prone and more realistic. Your test below should actually setup a DNS host like this test, but because autoHostRewrite() returns false, it won't be swapped.